exposure-notification/src/main/java/fi/thl/covid19/exposurenotification/efgs/FederationBatchSigner.java

```diff
@@ -0,0 +1,239 @@
+        }
+    }
+
+    private String sign(EfgsProto.DiagnosisKeyBatch data, KeyStore keyStore)
+            throws Exception {
+        PrivateKey key = (PrivateKey) keyStore.getKey(keyStoreKeyAlias, keyStorePassword);
+        X509Certificate cert = (X509Certificate) keyStore.getCertificate(keyStoreKeyAlias);
+        CMSSignedDataGenerator signedDataGenerator = new CMSSignedDataGenerator();
+        signedDataGenerator.addSignerInfoGenerator(createSignerInfo(cert, key));
+        signedDataGenerator.addCertificate(createCertificateHolder(cert));
+        CMSSignedData singedData = signedDataGenerator.generate(new CMSProcessableByteArray(generateBytesForSigning(data)), false);
+        return Base64.getEncoder().encodeToString(singedData.getEncoded());
+    }
+
+    private SignerInfoGenerator createSignerInfo(X509Certificate cert, PrivateKey key) throws OperatorCreationException,
+            CertificateEncodingException {
+        return new JcaSignerInfoGeneratorBuilder(createDigestBuilder()).build(createContentSigner(key), cert);
+    }
+
+    private X509CertificateHolder createCertificateHolder(X509Certificate cert) throws CertificateEncodingException,
+            IOException {
+        return new X509CertificateHolder(cert.getEncoded());
+    }
+
+    private DigestCalculatorProvider createDigestBuilder() throws OperatorCreationException {
+        return new JcaDigestCalculatorProviderBuilder().build();
+    }
+
+    private ContentSigner createContentSigner(PrivateKey privateKey) throws OperatorCreationException {
+        return new JcaContentSignerBuilder(DIGEST_ALGORITHM + privateKey.getAlgorithm()).build(privateKey);
+    }
+
+    private KeyStore initDevKeyStore() throws Exception {
+        KeyStore keyStore = KeyStore.getInstance("PKCS12");
+        keyStore.load(null);
+        KeyPair keyPair = generateKeyPair();
+        X509Certificate certificate = generateDevCertificate(keyPair);
+        keyStore.setCertificateEntry(keyStoreKeyAlias, certificate);
+        keyStore.setKeyEntry(keyStoreKeyAlias, keyPair.getPrivate(), keyStorePassword, new Certificate[]{certificate});
+
+        return keyStore;
+    }
+
+    private KeyPair generateKeyPair() throws NoSuchAlgorithmException {
+        KeyPairGenerator keyGen = KeyPairGenerator.getInstance("RSA");
+        keyGen.initialize(512);
+        return keyGen.generateKeyPair();
+    }
+
+    private X509Certificate generateDevCertificate(KeyPair keyPair) throws OperatorCreationException, IOException, CertificateException {
+        X500Name subject = new X500NameBuilder(BCStyle.INSTANCE)
+                .addRDN(BCStyle.C, "FI")
+                .addRDN(BCStyle.CN, "koronavilkku.dev")
+                .addRDN(BCStyle.O, "koronavilkku dev")
+                .build();
+
+        X509v3CertificateBuilder certBuilder = new X509v3CertificateBuilder(
+                subject,
+                BigInteger.valueOf(123456789),
+                Date.from(Instant.now()),
```

Comment:
They don't have an instant-based API? No-one should be using java.util.Date any more.

Reply:
dunno... this is latest I think.